docs/architecture/adr-0028-tracing.md

```diff
@@ -0,0 +1,96 @@
+despite that appearing to be the method definition that [I'm supposed to care
+about](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/ConcurrentEffect.scala#L42-L61).
+- The most obvious workaround is suggested by
+the `EitherT` concurrent (or `WriterT`, or take your
+pick), but the trait I'd need (`KleisliConcurrent`) is [private to
+`cats-effect`](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Concurrent.scala#L749).
+- Moreover, the [`natchez`
+example](https://github.com/tpolecat/natchez/blob/master/modules/examples/src/main/scala/Example.scala)
+has no idea about being a webservice -- it only knows how to ask a database for
+some information. The extra typeclass requirements from `BlazeServerBuilder`
+are a bit challenging to resolve.
+
+Rob Norris currently recommends not to use `natchez` in
+anything you need to work, and I think we should take his
+[advice](https://github.com/tpolecat/natchez/blame/efcefbf28c88ab977206df0f394d4141b3c5b2ca/README.md#L60).
+
+The (broken) implementation I wound up with is on
+[`feature/js/natchez`](https://github.com/jisantuc/tracing-demos/tree/feature/js/natchez).
+
+### `scala-opentracing`
+
+`scala-opentracing` ties its language very closely to the [Open Tracing
+specification](https://opentracing.io/specification/). It's a library used by
+Colisweb in production according to its readme. In summary, wherever you want
+to trace, you'll need a `TracingContextBuilder[F]` for your effect type. The
+only off-the-shelf non-trivial one is for DataDog. I don't know if there's a
+way to get a local copy of the DataDog service running (I don't think there
+is), so I pretended to have a fancier implementation that was really just
+wrapping a `new TracingContextBuilder[IO]` around `LoggingTracingContext[IO]`,
+since the `LoggingTracingContext` doesn't provide the implicit that I need. It
+was fairly easy to set up the dependency, with the exception of adding
+a repository to my resolvers (so we'd need to add that to Nexus in a few
+places as well). We could start with the `LoggingTracingContext` very easily.
+
+The out-of-the-box logging tracer [doesn't print or store tags
+anywhere](https://github.com/Colisweb/scala-opentracing/blob/e9563d6da8d921e1d4c25178ffb5131e047bd9b0/src/main/scala/com/colisweb/tracing/LoggingTracingContext.scala#L18),
+which is unfortunate, but I don't think it would be hard
+to roll our own, since it's an 86-line Scala file. If
+we want integration with a non-DataDog sink, I _think_
+it's the case that we can provide a `Tracer` from `io.opentracing` to the
+[`OpenTracingContext`](https://github.com/Colisweb/scala-opentracing/blob/e9563d6da8d921e1d4c25178ffb5131e047bd9b0/src/main/scala/com/colisweb/tracing/OpenTracingContext.scala)
+class and wrap it the same way I wrapped the `LoggingTracingContext`.
+
+DEPENDING ON TIME I MAY COME BACK AND TRY THAT but also http4s-tracer might
```

Comment:
This is looking pretty good - there are a few things working in favor of using `scala-opentracing` -- a few questions/things to maybe try out and see how hard they are.

[LoggingTracingContext](https://github.com/Colisweb/scala-opentracing/blob/0f55d1aab37ce175d6c6200f827113c6de52be2d/src/main/scala/com/colisweb/tracing/LoggingTracingContext.scala#L55-L56) it randomly generates the trace id that is propagated through child traces/spans; however, in our case we want to pull the value from a request header or randomly generate it. It looks like that is possible, but it might be good to prove that out before we decide if this is the approach or not.

There is some [work in progress](https://github.com/aws/aws-xray-sdk-java/issues/11) for adding OpenTracing support to AWS Xray - which makes me a little more comfortable with us starting out with `scala-opentracing`, starting with logs and then graduating to something else for where to push traces.

Reply:
So I answered my own question here after fooling around a little. Not sure that we'd use this implementation exactly, but here's what I did:

I overrode `wrapRoutes` in `TracedHttpRoutes` with the following:
```scala
def wrapHttpRoutes[F[_]: Sync](
      routes: Kleisli[OptionT[F, ?], TracedRequest[F], Response[F]],
      builder: TracingContextBuilder[F]
  ): HttpRoutes[F] = {
    Kleisli[OptionT[F, ?], Request[F], Response[F]] { req =>
      val operationName = "http4s-incoming-request"
      val traceHeader = req.headers.get(CaseInsensitiveString("X-Amzn-Trace-Id")) match {
        case Some(header) => header.value
        case _            => ""
      }
      val tags = Map(
        HTTP_METHOD.getKey -> req.method.name,
        HTTP_URL.getKey    -> req.uri.path.toString,
        "TraceId"          -> traceHeader
      )

      OptionT {
        builder(operationName, tags) use { context =>
          val tracedRequest = TracedRequest[F](req, context)
          val responseOptionWithTags = routes.run(tracedRequest) semiflatMap { response =>
            val tags = Map(
              HTTP_STATUS.getKey() -> response.status.code.toString
            ) ++
              response.headers.toList.map(h => (s"http.response.header.${h.name}" -> h.value)).toMap
            context
              .addTags(tags)
              .map(_ => response)
          }
          responseOptionWithTags.value
        }
      }
    }
```
The main difference is that it pulls out the trace header and puts it in `tags` so that the `builder` has access to it since type signature constraints seem to prevent passing in a request directly to the builder defined in your `Tracing` object. Then for `tracingContextBuilder` I passed in a custom `idGenerator` that checks if the `traceId` is an empty string, if it is not present then I generate a random `uuid`.

```scala
def tracingContextBuilder(implicit contextShift: ContextShift[IO],
                            timer: Timer[IO]): TracingContextBuilder[IO] =
    new TracingContextBuilder[IO] {

      def apply(operationName: String,
                tags: Map[String, String] = Map.empty): TracingContextResource[IO] = {

        val idGenerator: IO[String] = IO(tags.getOrElse("TraceId", s"${UUID.randomUUID()}") match {
          case "" => s"${UUID.randomUUID()}"
          case s  => s
        })

        LoggingTracingContext[IO](idGenerator = Some(idGenerator))(operationName, tags)
      }
    }
```

When making the following `http` call `http :8080/api/interpreter seed==1234 ops==12 X-Amzn-Trace-Id:CoolMan` I end up with the following in the logs

```
[info] [blaze-selector-0] INFO  o.h.b.c.n.NIO1SocketServerGroup - Accepted connection from /127.0.0.1:32914 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Starting Span CoolMan (http4s-incoming-request) 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Starting Span 5d1e4506-816b-4bf1-909f-5f8047a793ea (collect-ops) 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Finished Span 5d1e4506-816b-4bf1-909f-5f8047a793ea (collect-ops) in 2ms 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Starting Span 17bf7734-6426-4348-bae5-018616422572 (build-op) 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Finished Span 17bf7734-6426-4348-bae5-018616422572 (build-op) in 0ms 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Starting Span 96f26297-59f2-488a-9490-ffafdbf2da01 (send-response) 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Finished Span 96f26297-59f2-488a-9490-ffafdbf2da01 (send-response) in 1ms 
[info] [scala-execution-context-global-20] TRACE c.c.t.LoggingTracingContext$ - Trace CoolMan Finished Span CoolMan (http4s-incoming-request) in 4ms 
```

Reply:
This looks good to me and I think I know where all of these things go. I'll include another commit with "how to get whatever trace id you want".